packages/playground/src/services.ts

```diff
@@ -17,65 +18,116 @@ export function attachServices(host: BrowserHost) {
-    return TextDocument.create(model.uri.toString(), "cadl", 1, model.getValue());
+    return TextDocument.create(
+      model.uri.toString(),
+      "cadl",
+      model.getVersionId(),
+      model.getValue()
+    );
+  }
+
+  function lspArgs(model: monaco.editor.ITextModel, pos: monaco.Position) {
+    return {
+      textDocument: textDocumentForModel(model),
+      position: lspPosition(pos),
+    };
+  }
+
+  function lspPosition(pos: monaco.Position): lsp.Position {
+    return {
+      line: pos.lineNumber - 1,
+      character: pos.column - 1,
+    };
+  }
+
+  function monacoLocation(loc: lsp.Location): monaco.languages.Location {
+    return {
+      uri: monaco.Uri.parse(loc.uri),
+      range: monacoRange(loc.range),
+    };
+  }
+
+  function monacoRange(range: lsp.Range): monaco.IRange {
+    return {
+      startColumn: range.start.character + 1,
+      startLineNumber: range.start.line + 1,
+      endColumn: range.end.character + 1,
+      endLineNumber: range.end.line + 1,
+    };
+  }
+
+  function monacoTextEdit(edit: lsp.TextEdit): monaco.languages.TextEdit {
+    return {
+      range: monacoRange(edit.range),
+      text: edit.newText,
+    };
+  }
+
+  function monacoWorkspaceEdit(edit: lsp.WorkspaceEdit): monaco.languages.WorkspaceEdit {
+    const edits: monaco.languages.WorkspaceTextEdit[] = [];
+    for (const [uri, changes] of Object.entries(edit.changes ?? {})) {
+      const resource = monaco.Uri.parse(uri);
+      for (const change of changes) {
+        edits.push({ resource, edit: monacoTextEdit(change) });
+      }
+    }
+    return { edits };
   }
 
   monaco.languages.registerDefinitionProvider("cadl", {
     async provideDefinition(model, position) {
-      const result = await serverLib.gotoDefinition({
-        position: {
-          line: position.lineNumber - 1,
-          character: position.column - 1,
-        },
-        textDocument: textDocumentForModel(model),
-      });
-
-      if (!result) return;
-
-      const loc: monaco.languages.Location = {
-        uri: monaco.Uri.parse(result.uri),
-        range: {
-          startColumn: result.range.start.character + 1,
-          startLineNumber: result.range.start.line + 1,
-          endColumn: result.range.end.character + 1,
-          endLineNumber: result.range.end.line + 1,
-        },
-      };
+      const results = await serverLib.gotoDefinition(lspArgs(model, position));
+      return results.map(monacoLocation);
+    },
+  });
 
-      return loc;
+  monaco.languages.registerReferenceProvider("cadl", {
+    async provideReferences(model, position, context) {
+      const results = await serverLib.findReferences({ ...lspArgs(model, position), context });
+      return results.map(monacoLocation);
     },
   });
 
-  const lsConfig = serverLib.initialize({
-    capabilities: {},
-    processId: 1,
-    workspaceFolders: [],
-    rootUri: "inmemory://",
+  monaco.languages.registerRenameProvider("cadl", {
```

Comment:
Does this work automatically with some default shortcuts or is it only available via a context menu?

Reply:
There are default shortcuts. F2 to rename works great.